# Post-processing CSV

In [1]:
# https://ipython.org/ipython-doc/3/config/extensions/autoreload.html
%load_ext autoreload
%autoreload 2

In [2]:
from datetime import datetime


import cv2
import numpy as np
import matplotlib.pylab as plt

%matplotlib inline

In [3]:
import sys
sys.path.append("../common/")

In [4]:
from data_utils import get_scalers, LABELS, TRAIN_IMAGE_IDS
from image_utils import get_image_data
from submission_utils import submission_iterator, write_shp_from_polygons, write_shp_from_csv, write_shp_from_mask
from submission_utils import get_scaled_polygons, get_polygons, compute_label_image, get_data_csv, get_image_ids
from submission_utils import rewrite_submission
from visu_utils import plt_st, display_img_1b

In [5]:
for i, l in enumerate(LABELS):
    print i, l

0 None
1 Buildings
2 Misc. Manmade structures
3 Road
4 Track
5 Trees
6 Crops
7 Waterway
8 Standing water
9 Vehicle Large
10 Vehicle Small


Compare submission's predictions raw and post-processed on trained data with ground truth

In [6]:
from np_metrics import score

Get common image ids from submission file and training image ids

In [7]:
def write_shp_pred_pp(image_id, csv_file):
    data_csv = get_data_csv(image_id, csv_file)
    assert data_csv is not None, "No data_csv"
    write_shp_from_csv("../results/shp/%s_new.shp" % image_id, data_csv)
    polygons = get_scaled_polygons(data_csv)
    y_pred_raw = compute_label_image(image_id, polygons)
    y_pred_pp = mask_postprocessing(y_pred_raw, class_pp_func_list)
    write_shp_from_mask("../results/shp/%s_pp.shp" % image_id, image_id, y_pred_pp)

In [8]:
def compare_true_pred_pp(image_id, csv_file, write_shp=True, visualize_labels=()):
    y_true = get_image_data(image_id, 'label')
    data_csv = get_data_csv(image_id, csv_file)
    assert data_csv is not None, "No data_csv"
    if write_shp:
        write_shp_from_csv("../results/shp/%s_new.shp" % image_id, data_csv)
    polygons = get_scaled_polygons(data_csv)    
    y_pred_raw = compute_label_image(image_id, polygons)
    
    h = min(y_true.shape[0], y_pred_raw.shape[0])
    w = min(y_true.shape[1], y_pred_raw.shape[1])
    s1 = score(y_true[:h,:w,:], y_pred_raw[:h,:w,:])    
    print "True vs Pred Raw: ", s1
    
    polygons = shape_postprocessing(polygons, class_shape_pp_func_list)
    y_pred_pp = compute_label_image(image_id, polygons)
    y_pred_pp = mask_postprocessing(y_pred_pp, class_pp_func_list)
    
    h = min(y_true.shape[0], y_pred_pp.shape[0])
    w = min(y_true.shape[1], y_pred_pp.shape[1])
    s2 = score(y_true[:h,:w,:], y_pred_pp[:h,:w,:])
    print "True vs Pred PP: ", s2
    if write_shp:
        write_shp_from_mask("../results/shp/%s_pp.shp" % image_id, image_id, y_pred_pp)
        
        
        
    if len(visualize_labels) > 0:
        ll = len(visualize_labels)
        for i in visualize_labels:
            if i % 2 == 0:
                plt_st(12, 6)
            plt.subplot(131)
            display_img_1b(y_true[:,:,i])
            plt.title("Truth")
            plt.subplot(132)
            display_img_1b(y_pred_raw[:,:,i])
            plt.title("Raw")
            plt.subplot(133)
            display_img_1b(y_pred_pp[:,:,i])
            plt.title("PP")
        
        
        
    return s1, s2

In [10]:
#CSV_FILE_NEW = "../results/sub_mix_20170304_2329.csv"
#CSV_FILE_NEW = "../results/submission_2017-03-05-23-26.csv"
CSV_FILE_NEW = "../results/union_201703071611.csv"
CSV_FILE_TRAIN = "../results/sub_model_1_20170304_1826_nan.csv"


In [11]:
from postprocessing_utils import mask_postprocessing
from postprocessing_utils import trees_postprocessing, buildings_postprocessing, roads_postprocessing, standing_water_postprocessing
from postprocessing_utils import shape_postprocessing, roads_shape_postprocessing
from postprocessing_utils import buildings_shape_postprocessing


class_pp_func_list = {
    1: buildings_postprocessing, 
#     3: roads_postprocessing,
    5: trees_postprocessing, 
    8: standing_water_postprocessing, 
}

class_shape_pp_func_list = {
    #1: buildings_shape_postprocessing, # Hand-made labeling is forbidden :(
    3: roads_shape_postprocessing
}

Check on all training data:

In [12]:
train_ids = TRAIN_IMAGE_IDS

In [13]:
total_s1 = 0
total_s2 = 0
for image_id in train_ids:
    print "--", image_id
    s1, s2 = compare_true_pred_pp(image_id, CSV_FILE_TRAIN, write_shp=False)
    total_s1 += s1
    total_s2 += s2

total_s1 *= 1.0 / len(train_ids)
total_s2 *= 1.0 / len(train_ids)
print "Final score: ", total_s1, total_s2

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124417
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.305632
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.336076
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.15652
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.136793
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.143905
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.364399
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.132173
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.406851
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.345627
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.240389
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.306018
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.204052
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.139813
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.249001

Standing water

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124523
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.302711
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.332018
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.155803
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.137345
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.145944
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.352157
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.131629
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.399476
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.338185
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.23354
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.295865
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.2014
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.139019
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.243999
-- 6110_3_1
True vs Pred Raw:  0.355562
True vs Pred PP:  0.358455
-- 6010_4_4
True vs Pred Raw:  0.120205
True vs Pred PP:  0.120205
-- 6170_2_4
True vs Pred Raw:  0.220104
True vs Pred PP:  0.219598
-- 6170_4_1
True vs Pred Raw:  0.117173
True vs Pred PP:  0.117173
-- 6170_0_4
True vs Pred Raw:  0.1282
True vs Pred PP:  0.1282
-- 6060_2_3
True vs Pred Raw:  0.183337
True vs Pred PP:  0.183337
-- 6070_2_3
True vs Pred Raw:  0.213583
True vs Pred PP:  0.212168
-- 6010_1_2
True vs Pred Raw:  0.119722
True vs Pred PP:  0.119722
-- 6040_4_4
True vs Pred Raw:  0.121834
True vs Pred PP:  0.121834
-- 6100_2_2
True vs Pred Raw:  0.256762
True vs Pred PP:  0.256735
Final score:  0.214040441513 0.214841640592



Buildings + Roads1 + Trees + Water

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124417
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.300468
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.333829
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.15652
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.136793
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.143904
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.36002
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.132173
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.406208
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.343092
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.23561
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.297894
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.204052
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.1398
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.246068
-- 6110_3_1
True vs Pred Raw:  0.355562
True vs Pred PP:  0.36527
-- 6010_4_4
True vs Pred Raw:  0.120205
True vs Pred PP:  0.120364
-- 6170_2_4
True vs Pred Raw:  0.220104
True vs Pred PP:  0.218764
-- 6170_4_1
True vs Pred Raw:  0.117173
True vs Pred PP:  0.117373
-- 6170_0_4
True vs Pred Raw:  0.1282
True vs Pred PP:  0.12805
-- 6060_2_3
True vs Pred Raw:  0.183337
True vs Pred PP:  0.185514
-- 6070_2_3
True vs Pred Raw:  0.213583
True vs Pred PP:  0.21341
-- 6010_1_2
True vs Pred Raw:  0.119722
True vs Pred PP:  0.120306
-- 6040_4_4
True vs Pred Raw:  0.121834
True vs Pred PP:  0.120114
-- 6100_2_2
True vs Pred Raw:  0.256762
True vs Pred PP:  0.262434
Final score:  0.214040441513 0.21649792999

Save new submission

In [13]:
submission_file = '../results/submission_' + str(datetime.now().strftime("%Y-%m-%d-%H-%M")) + '.csv'
rewrite_submission(CSV_FILE_NEW, submission_file, class_pp_func_list, class_shape_pp_func_list)

-- 6120_2_4 10 0
-- 6120_2_3 10 1


Self-intersection at or near point 3193.5167785234898 1985.6040268456377


-- 6120_2_1 10 2
-- 6180_2_4 10 3


Ring Self-intersection at or near point 2247 3101


-- 6180_2_1 10 4


Ring Self-intersection at or near point 1126 3165


-- 6180_2_0 10 5


Ring Self-intersection at or near point 1229 3266


-- 6180_2_3 10 6


Ring Self-intersection at or near point 1582 2914


-- 6180_2_2 10 7


Ring Self-intersection at or near point 2219 3108


-- 6180_0_3 10 8


Ring Self-intersection at or near point 255 3022


-- 6180_0_2 10 9


Ring Self-intersection at or near point 760 2884


-- 6180_0_1 10 10


Ring Self-intersection at or near point 2672 2043


-- 6180_0_0 10 11


Ring Self-intersection at or near point 591 2383


-- 6180_0_4 10 12


Ring Self-intersection at or near point 1726 2680
Ring Self-intersection at or near point 2538 369


-- 6080_4_4 10 13


Ring Self-intersection at or near point 2923 3030


-- 6080_4_2 10 14


Self-intersection at or near point 3210.7092084006463 700.74798061389333
Ring Self-intersection at or near point 3294 1265
Ring Self-intersection at or near point 3260 830


-- 6080_4_3 10 15


Ring Self-intersection at or near point 2257 2583


-- 6080_4_0 10 16


Ring Self-intersection at or near point 1778 601


-- 6080_4_1 10 17


Ring Self-intersection at or near point 2116 2642


-- 6090_4_1 10 18


Ring Self-intersection at or near point 907 1889


-- 6090_4_0 10 19
-- 6090_4_3 10 20


Ring Self-intersection at or near point 573 1487


-- 6090_4_2 10 21


Ring Self-intersection at or near point 974 2287


-- 6090_4_4 10 22


Ring Self-intersection at or near point 3310 3253


-- 6180_4_4 10 23


Ring Self-intersection at or near point 1993 3290


-- 6180_4_3 10 24


Ring Self-intersection at or near point 2182 2571


-- 6180_4_2 10 25


Ring Self-intersection at or near point 3078 2310


-- 6180_4_1 10 26


Ring Self-intersection at or near point 99 3299


-- 6180_4_0 10 27


Self-intersection at or near point 66.73684210526315 3135.2631578947367


-- 6160_3_2 10 28


Ring Self-intersection at or near point 980 1634


-- 6160_3_3 10 29


Ring Self-intersection at or near point 1978 247


-- 6160_3_0 10 30


Ring Self-intersection at or near point 1900 3029


-- 6160_3_1 10 31


Ring Self-intersection at or near point 321 2123


-- 6160_3_4 10 32


Ring Self-intersection at or near point 321 2981


-- 6080_2_4 10 33
-- 6080_2_0 10 34


Ring Self-intersection at or near point 3098 1396
Self-intersection at or near point 2138.0884210526315 1184.0147368421053


-- 6080_2_1 10 35


Ring Self-intersection at or near point 208 1267


-- 6080_2_2 10 36


Self-intersection at or near point 1731 3210
Self-intersection at or near point 2622 1644


-- 6080_2_3 10 37


Ring Self-intersection at or near point 2541 1297


-- 6080_0_2 10 38


Self-intersection at or near point 825.468085106383 1083.7872340425531


-- 6080_0_3 10 39


Ring Self-intersection at or near point 410 2778


-- 6080_0_0 10 40


Ring Self-intersection at or near point 3312 2626


-- 6080_0_1 10 41
-- 6080_0_4 10 42


Ring Self-intersection at or near point 3126 2813


-- 6010_0_4 10 43


Ring Self-intersection at or near point 3000 3182


-- 6010_0_1 10 44
-- 6010_0_0 10 45
-- 6010_0_3 10 46


Ring Self-intersection at or near point 2089 2854


-- 6010_0_2 10 47


Ring Self-intersection at or near point 2573 3135


-- 6010_2_3 10 48


Ring Self-intersection at or near point 550 2873


-- 6010_2_2 10 49


Ring Self-intersection at or near point 2222 1331


-- 6010_2_1 10 50
-- 6010_2_0 10 51


Ring Self-intersection at or near point 2347 2979


-- 6010_2_4 10 52


Ring Self-intersection at or near point 1145 2445
Ring Self-intersection at or near point 3098 419


-- 6010_4_1 10 53
-- 6170_3_4 10 54
-- 6010_4_3 10 55


Ring Self-intersection at or near point 3084 2760


-- 6170_3_1 10 56


Ring Self-intersection at or near point 2554 1390


-- 6170_3_0 10 57


Ring Self-intersection at or near point 2324 3302


-- 6170_3_3 10 58


Ring Self-intersection at or near point 102 3172


-- 6170_3_2 10 59


Ring Self-intersection at or near point 1282 1612


-- 6170_1_3 10 60


Ring Self-intersection at or near point 1274 2893


-- 6170_1_2 10 61


Ring Self-intersection at or near point 2640 757


-- 6170_1_1 10 62


Ring Self-intersection at or near point 486 1123


-- 6170_1_0 10 63


Ring Self-intersection at or near point 813 2200


-- 6170_1_4 10 64


Ring Self-intersection at or near point 278 2968


-- 6130_4_2 10 65
-- 6130_4_3 10 66


Ring Self-intersection at or near point 834 3163


-- 6130_4_0 10 67
-- 6130_4_1 10 68


Ring Self-intersection at or near point 3010 3195


-- 6130_4_4 10 69


Ring Self-intersection at or near point 644 120


-- 6150_3_3 10 70


Ring Self-intersection at or near point 2472 2090


-- 6150_3_2 10 71


Ring Self-intersection at or near point 1589 1482


-- 6150_3_1 10 72


Ring Self-intersection at or near point 1962 865


-- 6150_3_0 10 73


Ring Self-intersection at or near point 87 1835


-- 6150_3_4 10 74
-- 6130_2_0 10 75


Ring Self-intersection at or near point 1352 2040


-- 6130_2_1 10 76


Ring Self-intersection at or near point 1026 2413


-- 6130_2_2 10 77


Ring Self-intersection at or near point 16 1315


-- 6130_2_3 10 78
-- 6130_2_4 10 79
-- 6130_0_4 10 80


Ring Self-intersection at or near point 1842 232
Ring Self-intersection at or near point 133 415


-- 6130_0_2 10 81
-- 6130_0_3 10 82
-- 6130_0_0 10 83
-- 6130_0_1 10 84
-- 6150_1_4 10 85


Ring Self-intersection at or near point 1608 1794


-- 6150_1_1 10 86


Ring Self-intersection at or near point 1913 2894


-- 6150_1_0 10 87


Ring Self-intersection at or near point 220 3262


-- 6150_1_3 10 88
-- 6150_1_2 10 89
-- 6180_3_4 10 90


Ring Self-intersection at or near point 2912 2643


-- 6180_3_0 10 91


Ring Self-intersection at or near point 679 2913


-- 6180_3_1 10 92


Ring Self-intersection at or near point 1616 3315


-- 6180_3_2 10 93


Ring Self-intersection at or near point 641 2489


-- 6180_3_3 10 94


Ring Self-intersection at or near point 2773 2417


-- 6180_1_2 10 95


Ring Self-intersection at or near point 1318 3074


-- 6180_1_3 10 96


Ring Self-intersection at or near point 951 3177


-- 6180_1_0 10 97


Ring Self-intersection at or near point 1270 3117


-- 6180_1_1 10 98


Ring Self-intersection at or near point 739 2749


-- 6180_1_4 10 99


Ring Self-intersection at or near point 803 3186


-- 6010_1_4 10 100


Self-intersection at or near point 1260.3214285714287 1972.6785714285713


-- 6010_1_0 10 101


Ring Self-intersection at or near point 3388 3036


-- 6010_1_1 10 102


Ring Self-intersection at or near point 18 2821


-- 6010_1_2 10 103


Ring Self-intersection at or near point 142 1651


-- 6010_1_3 10 104


Ring Self-intersection at or near point 1481 3052


-- 6010_3_2 10 105


Ring Self-intersection at or near point 3319 3003


-- 6010_3_3 10 106


Ring Self-intersection at or near point 1051 3235


-- 6010_3_0 10 107


Ring Self-intersection at or near point 2876 2216


-- 6010_3_1 10 108
-- 6010_3_4 10 109


Ring Self-intersection at or near point 2329 3223
Self-intersection at or near point 2183.2452830188681 1021.2075471698113


-- 6020_0_4 10 110


Ring Self-intersection at or near point 1337 2690


-- 6020_0_0 10 111


Ring Self-intersection at or near point 602 2275


-- 6020_0_1 10 112
-- 6020_0_2 10 113
-- 6020_0_3 10 114


Ring Self-intersection at or near point 2577 1498


-- 6020_2_2 10 115


Ring Self-intersection at or near point 6 1284


-- 6020_2_3 10 116
-- 6020_2_0 10 117
-- 6020_2_1 10 118
-- 6020_2_4 10 119
-- 6020_4_0 10 120
-- 6020_4_1 10 121
-- 6020_4_2 10 122


Ring Self-intersection at or near point 2464 1586


-- 6020_4_3 10 123
-- 6020_4_4 10 124
-- 6110_0_1 10 125
-- 6150_0_4 10 126


Ring Self-intersection at or near point 599 1405


-- 6150_0_0 10 127
-- 6150_0_1 10 128


Ring Self-intersection at or near point 2176 3046


-- 6150_0_2 10 129
-- 6150_0_3 10 130


Ring Self-intersection at or near point 697 1605


-- 6100_2_1 10 131
-- 6100_2_0 10 132
-- 6150_2_2 10 133


Ring Self-intersection at or near point 1563 2004


-- 6150_2_0 10 134


Ring Self-intersection at or near point 2504 2447


-- 6150_2_1 10 135


Ring Self-intersection at or near point 990 347


-- 6150_2_4 10 136


Ring Self-intersection at or near point 670 2763


-- 6170_4_2 10 137


Ring Self-intersection at or near point 932 664


-- 6050_3_1 10 138
-- 6170_4_3 10 139
-- 6150_4_0 10 140
-- 6150_4_1 10 141


Ring Self-intersection at or near point 226 2369


-- 6150_4_2 10 142
-- 6150_4_3 10 143
-- 6150_4_4 10 144
-- 6170_4_0 10 145


Ring Self-intersection at or near point 537 2179


-- 6070_3_4 10 146


Ring Self-intersection at or near point 2446 2732


-- 6070_3_0 10 147
-- 6070_3_1 10 148
-- 6070_3_2 10 149


Ring Self-intersection at or near point 103 3074


-- 6070_3_3 10 150


Ring Self-intersection at or near point 182 2289


-- 6070_1_2 10 151


Ring Self-intersection at or near point 2247 2893


-- 6070_1_3 10 152


Self-intersection at or near point 1523.1354166666667 1383.40625


-- 6070_1_0 10 153
-- 6070_1_1 10 154
-- 6070_1_4 10 155
-- 6120_0_3 10 156


Self-intersection at or near point 2893.91796875 1412.27734375


-- 6110_1_1 10 157
-- 6110_1_0 10 158
-- 6110_1_3 10 159
-- 6110_1_4 10 160
-- 6110_3_4 10 161
-- 6110_3_3 10 162


Ring Self-intersection at or near point 3351 2223


-- 6110_3_2 10 163


Self-intersection at or near point 2321.1505376344085 1795.8924731182797


-- 6110_3_0 10 164


Self-intersection at or near point 2033.4343434343434 1757.6363636363637


-- 6060_2_1 10 165
-- 6140_3_3 10 166


Ring Self-intersection at or near point 2417 737


-- 6020_1_4 10 167


Self-intersection at or near point 2473.032710280374 408.93925233644859
Ring Self-intersection at or near point 2379 2537


-- 6020_1_1 10 168


Ring Self-intersection at or near point 3319 2517


-- 6020_1_0 10 169


Ring Self-intersection at or near point 2142 1577


-- 6020_1_3 10 170
-- 6020_1_2 10 171
-- 6020_3_3 10 172
-- 6020_3_2 10 173


Ring Self-intersection at or near point 212 779


-- 6020_3_1 10 174
-- 6020_3_0 10 175


Ring Self-intersection at or near point 1708 2474


-- 6020_3_4 10 176
-- 6050_2_4 10 177
-- 6050_2_3 10 178
-- 6050_2_2 10 179
-- 6050_2_1 10 180


Self-intersection at or near point 2262.0761421319799 1590.0913705583757


-- 6050_2_0 10 181
-- 6100_3_4 10 182
-- 6100_3_0 10 183
-- 6100_3_1 10 184
-- 6100_3_2 10 185
-- 6100_3_3 10 186


Self-intersection at or near point 1807.2 352.39999999999998
Self-intersection at or near point 2592.8695652173915 303.95652173913044


-- 6060_1_1 10 187
-- 6060_1_0 10 188
-- 6060_1_3 10 189
-- 6060_1_2 10 190
-- 6060_1_4 10 191


Ring Self-intersection at or near point 1560 1308
Ring Self-intersection at or near point 1503 300


-- 6060_3_4 10 192


Interior is disconnected at or near point 2530 261


-- 6060_3_3 10 193


Self-intersection at or near point 559 3332


-- 6060_3_2 10 194


Ring Self-intersection at or near point 748 2251


-- 6060_3_1 10 195


Ring Self-intersection at or near point 2939 3272


-- 6060_3_0 10 196


Ring Self-intersection at or near point 1363 2672


-- 6070_2_4 10 197


Self-intersection at or near point 2658 3175.5


-- 6070_2_1 10 198


Ring Self-intersection at or near point 2853 2541


-- 6070_2_0 10 199
-- 6070_2_3 10 200


Self-intersection at or near point 1920.041450777202 3175.0051813471505
Ring Self-intersection at or near point 2259 2003


-- 6070_2_2 10 201


Ring Self-intersection at or near point 2601 2854


-- 6040_2_0 10 202


Ring Self-intersection at or near point 2701 3215


-- 6040_2_1 10 203


Ring Self-intersection at or near point 2846 2798


-- 6040_2_3 10 204


Ring Self-intersection at or near point 1507 536


-- 6040_2_4 10 205


Ring Self-intersection at or near point 2392 2209


-- 6070_0_3 10 206


Ring Self-intersection at or near point 2021 3221


-- 6070_0_2 10 207


Ring Self-intersection at or near point 1863 1801


-- 6070_0_1 10 208


Ring Self-intersection at or near point 3074 1116


-- 6070_0_0 10 209
-- 6070_0_4 10 210
-- 6100_4_3 10 211
-- 6040_0_4 10 212


Ring Self-intersection at or near point 1683 2650


-- 6040_0_2 10 213


Ring Self-intersection at or near point 711 2967


-- 6040_0_3 10 214


Ring Self-intersection at or near point 2563 2311


-- 6040_0_0 10 215


Ring Self-intersection at or near point 1898 2229


-- 6040_0_1 10 216


Ring Self-intersection at or near point 754 3272


-- 6050_0_1 10 217
-- 6050_0_0 10 218


Ring Self-intersection at or near point 695 2003


-- 6050_0_3 10 219
-- 6050_0_2 10 220
-- 6050_0_4 10 221
-- 6070_4_4 10 222
-- 6070_4_3 10 223


Ring Self-intersection at or near point 495 3293


-- 6070_4_2 10 224


Ring Self-intersection at or near point 168 1238


-- 6070_4_1 10 225


Self-intersection at or near point 1519.7142857142858 1029.4285714285713


-- 6070_4_0 10 226
-- 6140_3_0 10 227
-- 6140_3_2 10 228
-- 6100_4_4 10 229
-- 6140_3_4 10 230
-- 6100_4_2 10 231


Ring Self-intersection at or near point 3180 10


-- 6100_4_1 10 232
-- 6100_4_0 10 233
-- 6140_0_1 10 234
-- 6140_1_4 10 235
-- 6140_1_3 10 236


Ring Self-intersection at or near point 2900 2011


-- 6140_1_0 10 237


Self-intersection at or near point 2865 904


-- 6140_1_1 10 238


Ring Self-intersection at or near point 3215 1602


-- 6040_4_2 10 239


Ring Self-intersection at or near point 54 2287


-- 6040_4_3 10 240


Ring Self-intersection at or near point 1074 2397


-- 6040_4_0 10 241


Ring Self-intersection at or near point 2084 3317


-- 6040_4_1 10 242


Ring Self-intersection at or near point 2400 3193


-- 6040_4_4 10 243


Ring Self-intersection at or near point 964 2552


-- 6010_4_0 10 244


Ring Self-intersection at or near point 655 2549


-- 6110_0_0 10 245


Self-intersection at or near point 2168 1322


-- 6050_4_4 10 246
-- 6110_0_2 10 247
-- 6110_0_3 10 248
-- 6110_0_4 10 249
-- 6050_4_0 10 250
-- 6050_4_3 10 251
-- 6050_4_2 10 252
-- 6030_3_0 10 253


Ring Self-intersection at or near point 2256 275
Self-intersection at or near point 582.9859154929577 1575.1126760563379


-- 6030_3_1 10 254


Ring Self-intersection at or near point 3211 2270


-- 6030_3_2 10 255


Self-intersection at or near point 3349.8989898989898 270.97979797979798


-- 6030_3_3 10 256
-- 6030_3_4 10 257


Self-intersection at or near point 1697.4782608695652 805.56521739130437
Ring Self-intersection at or near point 1720 1018


-- 6110_2_4 10 258


Ring Self-intersection at or near point 2106 200


-- 6110_2_2 10 259


Self-intersection at or near point 1451.4666666666667 634


-- 6110_2_3 10 260
-- 6110_2_0 10 261
-- 6110_2_1 10 262


Ring Self-intersection at or near point 3376 268


-- 6120_4_1 10 263
-- 6120_4_0 10 264


Ring Self-intersection at or near point 2271 2133


-- 6120_4_3 10 265
-- 6120_4_2 10 266


Ring Self-intersection at or near point 1599 1638


-- 6110_4_4 10 267


Ring Self-intersection at or near point 2564 1480
Self-intersection at or near point 214.94594594594594 1373.2702702702702


-- 6110_4_1 10 268
-- 6110_4_2 10 269
-- 6110_4_3 10 270


Ring Self-intersection at or near point 2492 2205


-- 6060_4_4 10 271


Ring Self-intersection at or near point 1142 3157


-- 6060_4_0 10 272
-- 6060_4_1 10 273
-- 6060_4_2 10 274


Ring Self-intersection at or near point 2135 1399


-- 6060_4_3 10 275
-- 6100_2_4 10 276
-- 6050_3_4 10 277
-- 6050_3_2 10 278
-- 6050_3_3 10 279
-- 6050_3_0 10 280
-- 6100_2_2 10 281


Self-intersection at or near point 2396.0099009900991 168.9009900990099
Ring Self-intersection at or near point 3350 3212


-- 6100_0_3 10 282


Ring Self-intersection at or near point 1601 138
Ring Self-intersection at or near point 3308 3222


-- 6100_0_2 10 283


Ring Self-intersection at or near point 2070 1305


-- 6100_0_1 10 284


Ring Self-intersection at or near point 816 2062


-- 6100_0_0 10 285
-- 6100_0_4 10 286
-- 6060_0_0 10 287
-- 6060_0_1 10 288
-- 6060_0_2 10 289
-- 6060_0_3 10 290


Ring Self-intersection at or near point 2749 2174


-- 6060_0_4 10 291


Ring Self-intersection at or near point 1058 1329


-- 6160_1_0 10 292


Ring Self-intersection at or near point 3017 3121


-- 6060_2_4 10 293
-- 6060_2_2 10 294
-- 6060_2_0 10 295
-- 6160_1_1 10 296


Ring Self-intersection at or near point 3261 2567


-- 6040_3_1 10 297


Ring Self-intersection at or near point 2613 1777


-- 6040_3_0 10 298


Ring Self-intersection at or near point 2796 3188


-- 6040_3_3 10 299


Ring Self-intersection at or near point 728 2342


-- 6040_3_2 10 300


Ring Self-intersection at or near point 478 2779


-- 6040_3_4 10 301


Ring Self-intersection at or near point 3241 2657


-- 6090_0_4 10 302


Ring Self-intersection at or near point 1244 1809


-- 6160_1_3 10 303


Ring Self-intersection at or near point 1300 773


-- 6090_0_1 10 304


Ring Self-intersection at or near point 206 3089


-- 6090_0_0 10 305


Ring Self-intersection at or near point 2630 2825


-- 6090_0_3 10 306


Ring Self-intersection at or near point 812 2408


-- 6090_0_2 10 307


Ring Self-intersection at or near point 2176 1037


-- 6090_2_3 10 308


Ring Self-intersection at or near point 1079 1474


-- 6090_2_2 10 309


Ring Self-intersection at or near point 2662 320


-- 6090_2_1 10 310


Ring Self-intersection at or near point 2133 3238


-- 6040_1_4 10 311


Ring Self-intersection at or near point 697 3144


-- 6040_1_2 10 312


Ring Self-intersection at or near point 2246 2205


-- 6040_1_1 10 313


Ring Self-intersection at or near point 937 2945


-- 6090_2_4 10 314


Ring Self-intersection at or near point 2284 2698


-- 6030_0_4 10 315


Self-intersection at or near point 2464.1568627450979 132.80392156862746


-- 6030_0_3 10 316
-- 6030_0_2 10 317


Ring Self-intersection at or near point 1764 2733


-- 6030_0_1 10 318


Ring Self-intersection at or near point 1599 2637


-- 6030_0_0 10 319


Ring Self-intersection at or near point 156 2603


-- 6050_1_0 10 320
-- 6050_1_1 10 321
-- 6050_1_2 10 322
-- 6050_1_3 10 323
-- 6050_1_4 10 324
-- 6120_1_0 10 325
-- 6120_1_1 10 326


Ring Self-intersection at or near point 1754 695


-- 6120_1_2 10 327


Ring Self-intersection at or near point 1225 1597


-- 6120_1_3 10 328
-- 6120_1_4 10 329
-- 6120_0_4 10 330
-- 6160_1_4 10 331


Ring Self-intersection at or near point 2969 2163
Ring Self-intersection at or near point 984 1493


-- 6130_3_0 10 332


Ring Self-intersection at or near point 1613 3209


-- 6140_0_4 10 333
-- 6140_0_3 10 334
-- 6140_0_2 10 335
-- 6160_1_2 10 336


Ring Self-intersection at or near point 2641 3065


-- 6140_0_0 10 337


Ring Self-intersection at or near point 1420 83


-- 6120_3_4 10 338
-- 6120_3_2 10 339


Ring Self-intersection at or near point 425 2358


-- 6120_3_3 10 340
-- 6120_3_0 10 341


Ring Self-intersection at or near point 45 157


-- 6120_3_1 10 342


Ring Self-intersection at or near point 1730 1960


-- 6100_1_2 10 343


Self-intersection at or near point 2267 2032
Ring Self-intersection at or near point 714 2164


-- 6100_1_0 10 344


Ring Self-intersection at or near point 1399 3029


-- 6100_1_1 10 345
-- 6100_1_4 10 346
-- 6140_4_3 10 347
-- 6140_4_2 10 348


Self-intersection at or near point 1142.7692307692307 179.15384615384616


-- 6140_4_1 10 349


Self-intersection at or near point 3311.775212636695 1496.2515188335358


-- 6140_4_0 10 350
-- 6140_4_4 10 351


Ring Self-intersection at or near point 763 811


-- 6160_0_4 10 352


Ring Self-intersection at or near point 1840 2402


-- 6140_2_1 10 353
-- 6140_2_0 10 354


Ring Self-intersection at or near point 1553 599


-- 6140_2_3 10 355
-- 6140_2_2 10 356
-- 6140_2_4 10 357


Ring Self-intersection at or near point 2900 3052
Self-intersection at or near point 832.5212121212121 1340.5878787878787


-- 6030_4_3 10 358
-- 6030_4_2 10 359


Self-intersection at or near point 2540.0476190476193 2740.8095238095239
Self-intersection at or near point 2780.5094339622642 997.05660377358492


-- 6030_4_1 10 360


Ring Self-intersection at or near point 2393 2467
Self-intersection at or near point 2486 1411


-- 6030_4_0 10 361


Ring Self-intersection at or near point 129 2441


-- 6030_4_4 10 362
-- 6130_3_4 10 363
-- 6030_2_1 10 364


Ring Self-intersection at or near point 3386 761


-- 6030_2_0 10 365


Ring Self-intersection at or near point 1519 3093
Self-intersection at or near point 2792.1333333333332 1825.3333333333333


-- 6030_2_3 10 366


Ring Self-intersection at or near point 1178 725


-- 6030_2_2 10 367


Ring Self-intersection at or near point 2750 2942


-- 6030_2_4 10 368


Self-intersection at or near point 2430.9200000000001 3313.6399999999999


-- 6090_3_2 10 369
-- 6090_3_3 10 370


Ring Self-intersection at or near point 742 2622


-- 6090_3_0 10 371


Ring Self-intersection at or near point 434 2543


-- 6090_3_1 10 372


Ring Self-intersection at or near point 782 2596


-- 6090_3_4 10 373


Ring Self-intersection at or near point 501 2603


-- 6080_3_4 10 374
-- 6170_4_4 10 375


Ring Self-intersection at or near point 3086 3215


-- 6080_3_1 10 376


Ring Self-intersection at or near point 1598 987


-- 6080_3_0 10 377


Self-intersection at or near point 1572.8141112618725 2502.5603799185888


-- 6080_3_3 10 378
-- 6080_3_2 10 379


Self-intersection at or near point 2857.0384615384614 1959


-- 6160_4_1 10 380


Ring Self-intersection at or near point 858 2924


-- 6160_4_0 10 381


Ring Self-intersection at or near point 1955 3091


-- 6160_4_3 10 382


Ring Self-intersection at or near point 3196 3109


-- 6160_4_2 10 383


Ring Self-intersection at or near point 1387 1383


-- 6160_4_4 10 384


Ring Self-intersection at or near point 751 2703
Ring Self-intersection at or near point 3222 2833


-- 6080_1_3 10 385
-- 6080_1_2 10 386


Ring Self-intersection at or near point 1136 1087


-- 6080_1_1 10 387
-- 6080_1_0 10 388
-- 6080_1_4 10 389
-- 6170_2_0 10 390


Ring Self-intersection at or near point 1569 2231


-- 6170_2_1 10 391


Ring Self-intersection at or near point 362 3211


-- 6170_2_2 10 392


Ring Self-intersection at or near point 2806 2833


-- 6170_2_3 10 393


Ring Self-intersection at or near point 1708 864


-- 6170_0_2 10 394


Ring Self-intersection at or near point 3320 2846


-- 6170_0_3 10 395


Ring Self-intersection at or near point 1206 2879


-- 6170_0_0 10 396


Ring Self-intersection at or near point 1988 3244


-- 6170_0_1 10 397


Ring Self-intersection at or near point 2859 2650


-- 6090_1_4 10 398
-- 6120_4_4 10 399
-- 6090_1_0 10 400


Ring Self-intersection at or near point 1326 3022


-- 6090_1_1 10 401


Ring Self-intersection at or near point 2010 2698


-- 6090_1_2 10 402
-- 6090_1_3 10 403
-- 6120_0_1 10 404
-- 6120_0_0 10 405
-- 6030_1_4 10 406


Ring Self-intersection at or near point 3382 100


-- 6120_0_2 10 407


Ring Self-intersection at or near point 1103 2035


-- 6030_1_2 10 408


Ring Self-intersection at or near point 3225 184


-- 6030_1_3 10 409
-- 6030_1_0 10 410


Ring Self-intersection at or near point 3244 2929


-- 6030_1_1 10 411


Ring Self-intersection at or near point 907 2089


-- 6130_3_1 10 412


Ring Self-intersection at or near point 882 976


-- 6050_4_1 10 413
-- 6130_3_3 10 414


Ring Self-intersection at or near point 561 136


-- 6130_3_2 10 415
-- 6160_0_1 10 416


Ring Self-intersection at or near point 2771 3335


-- 6160_0_0 10 417


Ring Self-intersection at or near point 3280 1710


-- 6160_0_3 10 418


Ring Self-intersection at or near point 883 2987


-- 6160_0_2 10 419


Ring Self-intersection at or near point 400 3158


-- 6160_2_3 10 420


Ring Self-intersection at or near point 781 3105


-- 6160_2_2 10 421


Ring Self-intersection at or near point 2078 1248


-- 6160_2_0 10 422
-- 6160_2_4 10 423


Ring Self-intersection at or near point 979 1158


-- 6130_1_4 10 424


Self-intersection at or near point 2758.0235294117647 782.88235294117646
Ring Self-intersection at or near point 1744 2340


-- 6130_1_3 10 425


Ring Self-intersection at or near point 3203 1515


-- 6130_1_2 10 426
-- 6130_1_1 10 427


Ring Self-intersection at or near point 1020 2057


-- 6130_1_0 10 428


In [14]:
import pandas as pd

In [15]:
df1 = pd.read_csv(CSV_FILE_NEW)
df2 = pd.read_csv(submission_file)

In [16]:
assert (df1['ImageId'] == df2['ImageId']).all(), "WTF"
assert (df1['ClassType'] == df2['ClassType']).all(), "WTF"

In [17]:
# assert (df1['MultipolygonWKT'] == df2['MultipolygonWKT']).all(), "WTF"

## Visual data analysis

### Generate a 6XYZ image and SHP

In [19]:
import gdal
from visu_utils import scale_percentile

def generate_complete_image(image_id):
    filename = os.path.join("..", "results", "visu", '%s.tif' % image_id)
    if os.path.exists(filename):
        print "Image is already created. Remove it"
        os.remove(filename)        

    tile_size = (1024, 1024)
    n, m = 5, 5
    complete_image = np.zeros((5*tile_size[0], 5*tile_size[1]), dtype=np.uint16)
    counter = 0
    for i in range(n):
        ys = i*tile_size[1]
        ye = ys + tile_size[1]
        for j in range(m):
            xs = j*tile_size[0]
            xe = xs + tile_size[0]
            _id = image_id + "_%i_%i" % (i, j)
            img = get_image_data(_id, 'pan')
            img = cv2.resize(img, dsize=tile_size)    
            complete_image[ys:ye, xs:xe] = img

    complete_image_8u = scale_percentile(complete_image, 0.1, 99.9)             
    # Initialize output file:
    driver = gdal.GetDriverByName('GTiff')
    dataset = driver.Create(filename, complete_image_8u.shape[1], complete_image_8u.shape[0], 1, gdal.GDT_Byte)
    assert dataset is not None, "Failed to create a file"

    band = dataset.GetRasterBand(1)
    band.WriteArray(complete_image_8u, xoff=0, yoff=0)

    dataset = None

In [24]:
import os
from submission_utils import get_data_csv
from shapely.wkt import loads
from shapely.affinity import scale, translate

def get_transformed_polygons(data_csv, x, y, w, h):
    out = {}
    image_id = data_csv[0][0]
    xs, ys = get_scalers(image_id, h, w)
    for poly_info in data_csv:
        if "MULTIPOLYGON" not in poly_info[2][:20]:
            continue
        polygons = loads(poly_info[2])
        new_polygons = scale(polygons, xfact=xs, yfact=ys, origin=(0, 0, 0))        
        new_polygons = translate(new_polygons, xoff=x, yoff=y)
        out[int(poly_info[1])] = new_polygons
    return out

from collections import defaultdict

def generate_complete_shp_from_csv(image_id, csv_filename, suffix=''):
    
    filename = os.path.join("..", "results", "visu", '%s_%s.shp' % (image_id, suffix))
    if os.path.exists(filename):
        print "Image is already created. Remove it"
        os.remove(filename)        
        
    tile_size = (1024, 1024)
    n, m = 5, 5
    all_polygons = defaultdict(list)
    counter = 0
    for i in range(n):
        y = i*tile_size[1]
        for j in range(m):
            x = j*tile_size[0]
            _id = image_id + "_%i_%i" % (i, j)
            print "--", _id
            data_csv = get_data_csv(_id, csv_filename)        
            if data_csv is None:
                continue
            polygons = get_transformed_polygons(data_csv, x, y, tile_size[0], tile_size[1])
            for k in polygons:
                all_polygons[k].append(polygons[k])
    write_shp_from_polygons(filename, image_id, all_polygons)


In [14]:
CSV_FILE_LAST = submission_file
CSV_FILE_LAST[-20:-4]

In [22]:
# get_image_ids(CSV_FILE_LAST)

In [61]:
image_id = "6110"
# image_id = "6120"
# image_id = "6070"
# image_id = "6080"
# image_id = "6050"

# image_id = "6140"

In [46]:
generate_complete_image(image_id)

In [62]:
generate_complete_shp_from_csv(image_id, CSV_FILE_LAST, CSV_FILE_LAST[-20:-4])

-- 6110_0_0
-- 6110_0_1
-- 6110_0_2
-- 6110_0_3
-- 6110_0_4
-- 6110_1_0
-- 6110_1_1
-- 6110_1_2
-- 6110_1_3
-- 6110_1_4
-- 6110_2_0
-- 6110_2_1
-- 6110_2_2
-- 6110_2_3
-- 6110_2_4
-- 6110_3_0
-- 6110_3_1
-- 6110_3_2
-- 6110_3_3
-- 6110_3_4
-- 6110_4_0
-- 6110_4_1
-- 6110_4_2
-- 6110_4_3
-- 6110_4_4
Written succesfully file :  ../results/visu/6110_2017-03-07-02-17.shp


### Generate a SHP with all prediction for all image_id

In [25]:
import os
from collections import defaultdict
from data_utils import ALL_IMAGE_IDS
all_ids = sorted(np.unique([iid[:-4] for iid in ALL_IMAGE_IDS]).tolist())

def generate_all_shp_from_csv(csv_filename, suffix=''):
    filename = os.path.join("..", "results", "visu", 'all_predictions_%s.shp' % (suffix))
    assert not os.path.exists(filename), "Image is already created. Remove it manually"
        
    n_tiles = 18
    nx = 5
    ny = int(np.ceil(n_tiles / float(nx)))
    
    tile_size = (850, 850)
    all_polygons = defaultdict(list)
    count = 0
    for data_csv in submission_iterator(csv_filename):  
        image_id = data_csv[0][0]
        count += 1
        print "--", image_id
        index = all_ids.index(image_id[:4])
        i, j = (int(image_id[7]), int(image_id[5]))
        
        xi = index % nx * (5*tile_size[0]) + i*tile_size[0]
        yi = int(np.floor(index * 1.0 / nx)) * (5*tile_size[1]) + j*tile_size[1]        
        #print xi, yi
        polygons = get_transformed_polygons(data_csv, xi, yi, tile_size[0], tile_size[1])
        for k in polygons:
            all_polygons[k].append(polygons[k])
    write_shp_from_polygons(filename, "ALL", all_polygons)    

In [27]:
generate_all_shp_from_csv("../results/union_201703071611_pair_merge.csv", "union_201703071611_pair_merge")

-- 6120_2_4
-- 6120_2_3
-- 6120_2_1
-- 6180_2_4
-- 6180_2_1
-- 6180_2_0
-- 6180_2_3
-- 6180_2_2
-- 6180_0_3
-- 6180_0_2
-- 6180_0_1
-- 6180_0_0
-- 6180_0_4
-- 6080_4_4
-- 6080_4_2
-- 6080_4_3
-- 6080_4_0
-- 6080_4_1
-- 6090_4_1
-- 6090_4_0
-- 6090_4_3
-- 6090_4_2
-- 6090_4_4
-- 6180_4_4
-- 6180_4_3
-- 6180_4_2
-- 6180_4_1
-- 6180_4_0
-- 6160_3_2
-- 6160_3_3
-- 6160_3_0
-- 6160_3_1
-- 6160_3_4
-- 6080_2_4
-- 6080_2_0
-- 6080_2_1
-- 6080_2_2
-- 6080_2_3
-- 6080_0_2
-- 6080_0_3
-- 6080_0_0
-- 6080_0_1
-- 6080_0_4
-- 6010_0_4
-- 6010_0_1
-- 6010_0_0
-- 6010_0_3
-- 6010_0_2
-- 6010_2_3
-- 6010_2_2
-- 6010_2_1
-- 6010_2_0
-- 6010_2_4
-- 6010_4_1
-- 6170_3_4
-- 6010_4_3
-- 6170_3_1
-- 6170_3_0
-- 6170_3_3
-- 6170_3_2
-- 6170_1_3
-- 6170_1_2
-- 6170_1_1
-- 6170_1_0
-- 6170_1_4
-- 6130_4_2
-- 6130_4_3
-- 6130_4_0
-- 6130_4_1
-- 6130_4_4
-- 6150_3_3
-- 6150_3_2
-- 6150_3_1
-- 6150_3_0
-- 6150_3_4
-- 6130_2_0
-- 6130_2_1
-- 6130_2_2
-- 6130_2_3
-- 6130_2_4
-- 6130_0_4
-- 6130_0_2
-- 6130_0_3
-- 6